### ISSUE TITLE
Import Data From Pandas DataFrame Data Type Issue

### ISSUE THREAD

**Issue by Protesuiq** [capitalone/giraffez#36]
When exporting data from Pandas that are not string-objects,  raises this error:

```
Connection: Connecting to data source 'db1' ...
Connection: Connection to 'db1' established successfully.
MLoad: Requesting column info for table 'DBC.TestTable' ...
MLoad: Initiating Teradata PT request (awaiting server)  ...
MLoad: Teradata PT request accepted.
Connection: Closing Teradata connection ...
MLoad: Closing Teradata PT connection ...
MLoad: Acquisition phase was not called before closing.
MLoad: Ending acquisition phase ...
MLoad: Acquisition phase ended.
MLoad: Teradata PT request complete.
Connection: Connection to 'db1' closed.
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-15-4282e9006bed> in <module>()
      3     rows = list(zip(*[df[i].values.tolist() for i in df]))
      4     for row in rows:
----> 5        mload.load_row(row)
      6     exit_code = mload.finish()
      7     print("finished load with code {0}".format(exit_code))

C:\Users\MYUSERNAME\AppData\Local\Continuum\Anaconda3\lib\site-packages\giraffez\mload.py in load_row(self, items, panic)
    372             self._initiate()
    373         try:
--> 374             data = self.processor(items)
    375             row_status = self.load.put_row(data)
    376             if not self._handle_error(row_status):

C:\Users\MYUSERNAME\AppData\Local\Continuum\Anaconda3\lib\site-packages\giraffez\utils.py in _pipeline(x)
     66 def pipeline(funcs):
     67     def _pipeline(x):
---> 68         return reduce(lambda a, f: f(a), funcs, x)
     69     return _pipeline
     70 

C:\Users\MYUSERNAME\AppData\Local\Continuum\Anaconda3\lib\site-packages\giraffez\utils.py in <lambda>(a, f)
     66 def pipeline(funcs):
     67     def _pipeline(x):
---> 68         return reduce(lambda a, f: f(a), funcs, x)
     69     return _pipeline
     70 

C:\Users\MYUSERNAME\AppData\Local\Continuum\Anaconda3\lib\site-packages\giraffez\encoders.py in _encoder(items)
    278                 data.append(value)
    279             elif column.type in VAR_TYPES:
--> 280                 data.append(struct.pack("H", len(item)) + ensure_bytes(item))
    281             elif column.type in DATE_TYPES:
    282                 date_value = Date.from_string(item)

TypeError: object of type 'float' has no len() 
```
Non-string object have no len(): I had to transform my data to str to bypass this issue.
```
for i in df.columns.tolist():
    df[i] = df[i].values.astype(str)
```

before passing it to mload:
```
with giraffez.MLoad("DBC.TestTable") as mload:
    mload.columns = df.columns.tolist()
    rows = list(zip(*[df[i].values.tolist() for i in df]))
    for row in rows:
       mload.load_row(row)
    exit_code = mload.finish()
    print("finished load with code {0}".format(exit_code))
    if exit_code == 0:
       mload.cleanup()
```

Is this a known issue?

**Comment by ChrisRx** [comment]
Data has to be a string if it is being loaded into a Teradata column where the data type is one that must be a string (character data types).  What is the data from the dataframe that you are trying to load into a Teradata character type that isn't a Python string?  I ask because I see that it is a float type and it is trying to put that in a character data type in Teradata (you have to convert incompatible types to compatible ones to put into Teradata).

**Comment by Protesuiq** [comment]
I have converted them to match the definitions of the created table in Teradata. My data types are objects, np.float64, np.int64, np.bool.


My Teradata MLOAD script works without an issue with the  CSV generated by df.to_csv function.

**Comment by ChrisRx** [comment]
The mload script works because you are converting the types implicitly by using `df.to_csv` (which must convert all columns to strings in order for it to write that to a file).  So in either case you have to convert the data types to compatible Teradata data types.  For example, if you used `df.to_csv` to write a common-delimiter file (with header) it would operate just like mload when you use the command-line interface to load it `giraffez mload ...` (but that would be a really roundabout way to do that, this was just to help illustrate what is happening when you use `df.to_csv`).

A couple of the core principles of giraffez is to be fast/efficient and promote data quality (correctness), so automatically converting floats to strings is not something that it currently will do.  There might be a compelling reason to do so, but libraries like numpy/pandas or even just Python itself make it really easy to handle these types, so it is often more logical to convert the values only in situations where they need to be converted (rather than incur the computational cost of checking it everytime for everyone).  For instance, you could rewrite your script like this:  

```Python
def coerce_floats(df):
    """Coerce numpy floats to Python strings"""
    float_types = ('float16', 'float32', 'float64')
    for name in df.columns.tolist():
        if df[name].dtype in float_types:
            df[name] = df[name].values.astype(str)


with giraffez.MLoad("DBC.TestTable") as mload:
    mload.columns = df.columns.tolist()
    coerce_floats(df)
    for row in df.values.tolist():
       mload.load_row(row)
    exit_code = mload.finish()
    print("finished load with code {0}".format(exit_code))
    if exit_code == 0:
       mload.cleanup()
```

Pandas DataFrame types use numpy arrays/types so it gets to use the great type conversion features of numpy.  As you can see above it is easy to create a function like `coerce_floats` that runs only once and directs the DataFrame to convert columns of a particular type to any other type (so you could add say numpy integers to that as well if you wanted to auto convert them).

With that said, I do think at a minimum that the error message should be better, and in fact the newest version of giraffez that should be releasing very soon, will have a much better error message indicating how you can fix your script (saying something indicating the mismatch in types and what is expected).

**Comment by Protesuiq** [comment]
I tried that already. I actually changed all my data types to string.

```
for i in df.columns.tolist():
    df[i] = df[i].values.astype(str)
```
:) It would be nice to have a function to send Pandas DataFrame to Teradata with Giraffez, similar to export:

```
with giraffez.Export(encoding='dict') as export:
    export.query = "SQL query"
    df = pd.DataFrame(export.results())
```
